pkg/client/client.go

```diff
@@ -267,12 +447,245 @@ func (c *ImmuClient) ZAdd(setReader io.Reader, score float64, keyReader io.Reade
 }
 
-func (c *ImmuClient) HealthCheck() error {
+// SafeZAdd ...
+func (c *ImmuClient) SafeZAdd(ctx context.Context, set []byte, score float64, key []byte) (*VerifiedIndex, error) {
+	start := time.Now()
+	if !c.isConnected() {
+		return nil, ErrNotConnected
+	}
+
+	root, err := c.rootservice.GetRoot(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := &schema.SafeZAddOptions{
+		Zopts: &schema.ZAddOptions{
+			Set:   set,
+			Score: score,
+			Key:   key,
+		},
+		RootIndex: &schema.Index{
+			Index: root.Index,
+		},
+	}
+
+	var metadata runtime.ServerMetadata
+	result, err := c.serviceClient.SafeZAdd(
+		ctx,
+		opts,
+		grpc.Header(&metadata.HeaderMD), grpc.Trailer(&metadata.TrailerMD),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	key2, err := store.SetKey(key, set, score)
+	if err != nil {
+		return nil, err
+	}
+
+	// This guard ensures that result.Leaf is equal to the item's hash computed
+	// from request values. From now on, result.Leaf can be trusted.
+	item := schema.Item{
+		Key:   key2,
+		Value: key,
+		Index: result.Index,
+	}
+	if !bytes.Equal(item.Hash(), result.Leaf) {
+		return nil, errors.New("proof does not match the given item")
+	}
+
+	verified, err := c.verifyAndSetRoot(result, root)
+	if err != nil {
+		return nil, err
+	}
+
+	c.Logger.Debugf("safezadd finished in %s", time.Since(start))
+
+	return &VerifiedIndex{
+			Index:    result.Index,
+			Verified: verified,
+		},
+		nil
+}
+
+func writeSeek(w io.WriteSeeker, msg []byte, offset int64) (int64, error) {
+	buf := make([]byte, 4)
+	binary.LittleEndian.PutUint32(buf, uint32(len(msg)))
+	if _, err := w.Seek(offset, io.SeekStart); err != nil {
+		return offset, err
+	}
+	if _, err := w.Write(buf); err != nil {
+		return offset, err
+	}
+	if _, err := w.Seek(offset+int64(len(buf)), io.SeekStart); err != nil {
+		return offset, err
+	}
+	if _, err := w.Write(msg); err != nil {
+		return offset, err
+	}
+	return offset + int64(len(buf)) + int64(len(msg)), nil
+}
+
+func readSeek(r io.ReadSeeker, offset int64) ([]byte, int64, error) {
+	if _, err := r.Seek(offset, io.SeekStart); err != nil {
+		return nil, offset, err
+	}
+	buf := make([]byte, 4)
+	o1 := offset + int64(len(buf))
+	if _, err := io.ReadFull(r, buf); err != nil {
+		return nil, o1, err
+	}
+	if _, err := r.Seek(o1, io.SeekStart); err != nil {
+		return nil, o1, err
+	}
+	size := binary.LittleEndian.Uint32(buf)
+	msg := make([]byte, size)
+	o2 := o1 + int64(size)
+	if _, err := io.ReadFull(r, msg); err != nil {
+		return nil, o2, err
+	}
+	return msg, o2, nil
+}
+
+// Backup to be used from Immu CLI
+func (c *ImmuClient) Backup(ctx context.Context, writer io.WriteSeeker) (int64, error) {
+	start := time.Now()
+
+	var counter int64
+
+	if !c.isConnected() {
+		return counter, ErrNotConnected
+	}
+
+	bkpClient, err := c.serviceClient.Backup(ctx, &empty.Empty{})
+	if err != nil {
+		return counter, err
+	}
+
+	var offset int64
+	var errs []string
+	for {
+		kvList, err := bkpClient.Recv()
```

Comment:
You need to close stream.
Take a look at https://grpc.io/docs/tutorials/basic/go/
stream.SendAndClose